File: epoxy-processortest/src/test/java/com/airbnb/epoxy/ModelProcessorTest.java
package com.airbnb.epoxy;

import com.google.testing.compile.JavaFileObjects;

import org.junit.Test;

import javax.tools.JavaFileObject;

import static com.airbnb.epoxy.ProcessorTestUtils.assertGenerationError;
import static com.airbnb.epoxy.ProcessorTestUtils.checkFileCompiles;
import static com.airbnb.epoxy.ProcessorTestUtils.assertGeneration;
import static com.airbnb.epoxy.ProcessorTestUtils.processors;
import static com.google.common.truth.Truth.assert_;
import static com.google.testing.compile.JavaSourceSubjectFactory.javaSource;
import static junit.framework.Assert.assertTrue;

/**
 * These processor tests are in their own module since the processor module can't depend on the
 * android EpoxyAdapter library that contains the EpoxyModel.
 */

@SuppressWarnings("ResultOfMethodCallIgnored")
public class ModelProcessorTest {
  @Test
  public void testSimpleModel() {
    assertGeneration("BasicModelWithAttribute.java", "BasicModelWithAttribute_.java");
  }

  @Test
  public void testModelWithAllFieldTypes() {
    assertGeneration("ModelWithAllFieldTypes.java", "ModelWithAllFieldTypes_.java");
  }

  @Test
  public void testModelWithConstructors() {
    assertGeneration("ModelWithConstructors.java", "ModelWithConstructors_.java");
  }

  @Test
  public void testModelWithSuper() {
    assertGeneration("ModelWithSuper.java", "ModelWithSuper_.java");
  }

  @Test
  public void testModelWithFieldAnnotation() {
    assertGeneration("ModelWithFieldAnnotation.java", "ModelWithFieldAnnotation_.java");
  }

  @Test
  public void testModelWithSuperClass() {
    JavaFileObject model = JavaFileObjects
        .forResource(GuavaPatch.patchResource("ModelWithSuperAttributes.java"));

    JavaFileObject generatedModel = JavaFileObjects.forResource(GuavaPatch.patchResource("ModelWithSuperAttributes_.java"));
    JavaFileObject generatedSubClassModel =
        JavaFileObjects.forResource(GuavaPatch.patchResource("ModelWithSuperAttributes$SubModelWithSuperAttributes_.java"));

    assert_().about(javaSource())
        .that(model)
        .processedWith(processors())
        .compilesWithoutError()
        .and()
        .generatesSources(generatedModel, generatedSubClassModel);
  }

  @Test
  public void testModelWithType() {
    assertGeneration("ModelWithType.java", "ModelWithType_.java");
  }

  @Test
  public void testModelWithoutHash() {
    assertGeneration("ModelWithoutHash.java", "ModelWithoutHash_.java");
  }

  @Test
  public void testDoNotHashModel() {
    assertGeneration("ModelDoNotHash.java", "ModelDoNotHash_.java");
  }

  @Test
  public void testModelWithFinalAttribute() {
    assertGeneration("ModelWithFinalField.java", "ModelWithFinalField_.java");
  }

  @Test
  public void testModelWithStaticAttributeFails() {
    assertGenerationError("ModelWithStaticField.java", "static");
  }

  @Test
  public void testModelWithPrivateClassFails() {
    assertGenerationError("ModelWithPrivateInnerClass.java", "private classes");
  }

  @Test
  public void testModelWithFinalClassFails() {
    assertGenerationError("ModelWithFinalClass.java", "");
  }

  @Test
  public void testModelThatDoesNotExtendEpoxyModelFails() {
    assertGenerationError("ModelWithoutEpoxyExtension.java", "must extend");
  }

  @Test
  public void testModelAsInnerClassFails() {
    assertGenerationError("ModelAsInnerClass.java", "Nested classes");
  }

  @Test
  public void testModelWithIntDefAnnotation() {
    assertGeneration("ModelWithIntDef.java", "ModelWithIntDef_.java");
  }

  @Test
  public void testModelWithAnnotatedClass() {
    assertGeneration("ModelWithAnnotatedClass.java", "ModelWithAnnotatedClass_.java");
  }

  @Test
  public void testModelWithAbstractClass() {
    JavaFileObject model = JavaFileObjects
        .forResource(GuavaPatch.patchResource("ModelWithAbstractClass.java"));

    assert_().about(javaSource())
        .that(model)
        .processedWith(processors())
        .compilesWithoutError();

    // We don't generate subclasses if the model is abstract unless it has a class annotation.
    boolean modelNotGenerated;
    try {
      JavaFileObjects.forResource("non-sense? Any wrong path will generate IAE");
      modelNotGenerated = false;
    } catch (IllegalArgumentException e) {
      modelNotGenerated = true;
    }

    assertTrue(modelNotGenerated);
  }

  @Test
  public void testModelWithAbstractClassAndAnnotation() {
    assertGeneration("ModelWithAbstractClassAndAnnotation.java",
        "ModelWithAbstractClassAndAnnotation_.java");
  }

  @Test
  public void testModelWithAnnotatedClassAndSuperClass() {
    JavaFileObject model = JavaFileObjects
        .forResource(GuavaPatch.patchResource("ModelWithAnnotatedClassAndSuperAttributes.java"));

    JavaFileObject generatedModel = JavaFileObjects
        .forResource(GuavaPatch.patchResource("ModelWithAnnotatedClassAndSuperAttributes_.java"));
    JavaFileObject generatedSubClassModel =
        JavaFileObjects.forResource(GuavaPatch.patchResource(
                "ModelWithAnnotatedClassAndSuperAttributes$SubModel"
            + "WithAnnotatedClassAndSuperAttributes_.java"));

    assert_().about(javaSource())
        .that(model)
        .processedWith(processors())
        .compilesWithoutError()
        .and()
        .generatesSources(generatedModel, generatedSubClassModel);
  }

  @Test
  public void testModelWithoutSetter() {
    assertGeneration("ModelWithoutSetter.java", "ModelWithoutSetter_.java");
  }

  @Test
  public void testModelReturningClassType() {
    assertGeneration("ModelReturningClassType.java", "ModelReturningClassType_.java");
  }

  @Test
  public void testModelReturningClassTypeWithVarargs() {
    assertGeneration("ModelReturningClassTypeWithVarargs.java",
        "ModelReturningClassTypeWithVarargs_.java");
  }

  @Test
  public void testModelWithVarargsConstructors() {
    assertGeneration("ModelWithVarargsConstructors.java", "ModelWithVarargsConstructors_.java");
  }

  @Test
  public void testModelWithHolderGeneratesNewHolderMethod() {
    assertGeneration("AbstractModelWithHolder.java", "AbstractModelWithHolder_.java");
  }

  @Test
  public void testGenerateDefaultLayoutMethod() {
    assertGeneration("GenerateDefaultLayoutMethod.java", "GenerateDefaultLayoutMethod_.java");
  }

  @Test
  public void testGenerateDefaultLayoutMethodFailsIfLayoutNotSpecified() {
    assertGenerationError("GenerateDefaultLayoutMethodNoLayout.java",
        "Model must specify a valid layout resource");
  }

  @Test
  public void testGeneratedDefaultMethodWithLayoutSpecifiedInParent() {
    JavaFileObject model = JavaFileObjects
        .forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodParentLayout.java"));

    JavaFileObject generatedNoLayoutModel = JavaFileObjects
        .forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodParentLayout$NoLayout_.java"));
    JavaFileObject generatedWithLayoutModel =
        JavaFileObjects.forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodParentLayout$WithLayout_.java"));

    assert_().about(javaSource())
        .that(model)
        .processedWith(processors())
        .compilesWithoutError()
        .and()
        .generatesSources(generatedNoLayoutModel, generatedWithLayoutModel);
  }

  @Test
  public void testGeneratedDefaultMethodWithLayoutSpecifiedInNextParent() {
    JavaFileObject model = JavaFileObjects
        .forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodNextParentLayout.java"));

    JavaFileObject generatedNoLayoutModel = JavaFileObjects
        .forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodNextParentLayout$NoLayout_.java"));
    JavaFileObject generatedStillNoLayoutModel = JavaFileObjects
        .forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodNextParentLayout$StillNoLayout_.java"));
    JavaFileObject generatedWithLayoutModel =
        JavaFileObjects.forResource(GuavaPatch.patchResource("GenerateDefaultLayoutMethodNextParentLayout$WithLayout_.java"));

    assert_().about(javaSource())
        .that(model)
        .processedWith(processors())
        .compilesWithoutError()
        .and()
        .generatesSources(generatedNoLayoutModel, generatedStillNoLayoutModel,
            generatedWithLayoutModel);
  }

  @Test
  public void testGeneratedDefaultMethodWithLayoutFailsIfNotSpecifiedInParent() {
    assertGenerationError("GenerateDefaultLayoutMethodParentStillNoLayout.java",
        "Model must specify a valid layout resource");
  }

  @Test
  public void modelWithViewClickListener() {
    assertGeneration("ModelWithViewClickListener.java", "ModelWithViewClickListener_.java");
  }

  @Test
  public void modelWithViewClickLongListener() {
    JavaFileObject model = JavaFileObjects
        .forResource(GuavaPatch.patchResource("ModelWithViewLongClickListener.java"));

    JavaFileObject generatedNoLayoutModel = JavaFileObjects
        .forResource(GuavaPatch.patchResource("ModelWithViewLongClickListener_.java"));

    assert_().about(javaSource())
        .that(model)
        .processedWith(processors())
        .compilesWithoutError()
        .and()
        .generatesSources(generatedNoLayoutModel);
  }

  @Test
  public void modelWithCheckedChangeListener() {
    assertGeneration("ModelWithCheckedChangeListener.java", "ModelWithCheckedChangeListener_.java");
  }

  @Test
  public void testModelWithPrivateAttributeWithoutGetterAndSetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithoutGetterAndSetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithoutSetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithoutSetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithoutGetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithoutGetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithIsPrefixGetter() {
    checkFileCompiles("ModelWithPrivateFieldWithIsPrefixGetter.java");
  }

  @Test
  public void testModelWithPrivateAttributeWithPrivateGetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithPrivateGetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithStaticGetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithStaticGetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithGetterWithParamsFails() {
    assertGenerationError("ModelWithPrivateFieldWithGetterWithParams.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithPrivateSetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithPrivateSetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithStaticSetterFails() {
    assertGenerationError("ModelWithPrivateFieldWithStaticSetter.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithPrivateAttributeWithGetterWithoutParamsFails() {
    assertGenerationError("ModelWithPrivateFieldWithSettterWithoutParams.java",
        "private fields without proper getter and setter");
  }

  @Test
  public void testModelWithAllPrivateFieldTypes() {
    assertGeneration("ModelWithAllPrivateFieldTypes.java", "ModelWithAllPrivateFieldTypes_.java");
  }

  @Test
  public void modelWithViewPrivateClickListener() {
    assertGeneration("ModelWithPrivateViewClickListener.java",
        "ModelWithPrivateViewClickListener_.java");
  }

  @Test
  public void modelWithPrivateFieldWithSameAsFieldGetterAndSetterName() {
    assertGeneration("ModelWithPrivateFieldWithSameAsFieldGetterAndSetterName.java",
        "ModelWithPrivateFieldWithSameAsFieldGetterAndSetterName_.java");
  }

  @Test
  public void testDoNotUseInToStringModel() {
    assertGeneration("ModelDoNotUseInToString.java", "ModelDoNotUseInToString_.java");
  }

  @Test
  public void modelWithAnnotation() {
    assertGeneration("ModelWithAnnotation.java", "ModelWithAnnotation_.java");
  }

  @Test
  public void testModelBuilderInterface() {
    assertGeneration("ModelWithAllFieldTypes.java", "ModelWithAllFieldTypesBuilder.java");
  }

  @Test
  public void generatedEpoxyModelGroup() {
    assertGeneration("EpoxyModelGroupWithAnnotations.java", "EpoxyModelGroupWithAnnotations_.java");
  }
}

File: epoxy-processortest/src/test/java/com/airbnb/epoxy/ModelProcessorTest.kt
package com.airbnb.epoxy

import com.airbnb.epoxy.ProcessorTestUtils.assertGeneration
import com.airbnb.epoxy.ProcessorTestUtils.assertGenerationError
import com.airbnb.epoxy.ProcessorTestUtils.checkFileCompiles
import com.google.testing.compile.JavaFileObjects
import junit.framework.Assert
import org.junit.Test

/**
 * These processor tests are in their own module since the processor module can't depend on the
 * android EpoxyAdapter library that contains the EpoxyModel.
 */
class ModelProcessorTest {
    @Test
    fun testSimpleModel() {
        assertGeneration("BasicModelWithAttribute.java", "BasicModelWithAttribute_.java")
    }

    @Test
    fun testModelWithAllFieldTypes() {
        assertGeneration("ModelWithAllFieldTypes.java", "ModelWithAllFieldTypes_.java")
    }

    @Test
    fun testModelWithConstructors() {
        assertGeneration("ModelWithConstructors.java", "ModelWithConstructors_.java")
    }

    @Test
    fun testModelWithSuper() {
        assertGeneration("ModelWithSuper.java", "ModelWithSuper_.java")
    }

    @Test
    fun testModelWithFieldAnnotation() {
        assertGeneration("ModelWithFieldAnnotation.java", "ModelWithFieldAnnotation_.java")
    }

    @Test
    fun testModelWithSuperClass() {
        val model = JavaFileObjects.forResource("ModelWithSuperAttributes.java".patchResource())

        val generatedModel =
            JavaFileObjects.forResource("ModelWithSuperAttributes_.java".patchResource())

        val generatedSubClassModel =
            JavaFileObjects.forResource("ModelWithSuperAttributes\$SubModelWithSuperAttributes_.java".patchResource())

        assertGeneration(
            sources = listOf(model),
            generatedFileObjects = listOf(generatedModel, generatedSubClassModel)
        )
    }

    @Test
    fun testModelWithType() {
        assertGeneration(
            "ModelWithType.java",
            "ModelWithType_.java",
            // The generics in the model don't seem to work with KSP/Kotlin. It isn't a real/good
            // use case anyway, so don't care about supporting it.
            compilationMode = CompilationMode.JavaAP
        )
    }

    @Test
    fun testModelWithoutHash() {
        assertGeneration("ModelWithoutHash.java", "ModelWithoutHash_.java")
    }

    @Test
    fun testDoNotHashModel() {
        assertGeneration("ModelDoNotHash.java", "ModelDoNotHash_.java")
    }

    @Test
    fun testModelWithFinalAttribute() {
        assertGeneration("ModelWithFinalField.java", "ModelWithFinalField_.java")
    }

    @Test
    fun testModelWithStaticAttributeFails() {
        assertGenerationError("ModelWithStaticField.java", "static")
    }

    @Test
    fun testModelWithPrivateClassFails() {
        assertGenerationError(
            "ModelWithPrivateInnerClass.java",
            "private classes",
            // KSP bug prevents us from testing this case https://github.com/google/ksp/issues/622
            compilationMode = CompilationMode.JavaAP
        )
    }

    @Test
    fun testModelWithFinalClassFails() {
        assertGenerationError("ModelWithFinalClass.java", "")
    }

    @Test
    fun testModelThatDoesNotExtendEpoxyModelFails() {
        assertGenerationError("ModelWithoutEpoxyExtension.java", "must extend")
    }

    @Test
    fun testModelAsInnerClassFails() {
        assertGenerationError("ModelAsInnerClass.java", "Nested model classes must be static")
    }

    @Test
    fun testModelWithIntDefAnnotation() {
        assertGeneration(
            "ModelWithIntDef.java",
            "ModelWithIntDef_.java",
            // Throws NPE in KSP due to https://github.com/google/ksp/issues/622
            compilationMode = CompilationMode.JavaAP
        )
    }

    @Test
    fun testModelWithAnnotatedClass() {
        assertGeneration("ModelWithAnnotatedClass.java", "ModelWithAnnotatedClass_.java")
    }

    @Test
    fun testModelWithAbstractClass() {
        val model = JavaFileObjects
            .forResource("ModelWithAbstractClass.java".patchResource())

        assertGeneration(
            sources = listOf(model)
        )

        // We don't generate subclasses if the model is abstract unless it has a class annotation.
        val modelNotGenerated: Boolean = try {
            JavaFileObjects.forResource("non-sense? Any wrong path will generate IAE")
            false
        } catch (e: IllegalArgumentException) {
            true
        }

        Assert.assertTrue(modelNotGenerated)
    }

    @Test
    fun testModelWithAbstractClassAndAnnotation() {
        assertGeneration(
            "ModelWithAbstractClassAndAnnotation.java",
            "ModelWithAbstractClassAndAnnotation_.java"
        )
    }

    @Test
    fun testKotlinModel_kapt() {
        assertGeneration(
            "ModelProcessorTest/testKotlinModel/Model.kt",
            "ModelProcessorTest/testKotlinModel/Model_.java",
            compilationMode = CompilationMode.KAPT
        )
    }

    @Test
    fun testKotlinModel_ksp() {
        assertGeneration(
            "ModelProcessorTest/testKotlinModel/Model.kt",
            "ModelProcessorTest/testKotlinModel/Model_.java",
            compilationMode = CompilationMode.KSP
        )
    }

    @Test
    fun testModelWithAnnotatedClassAndSuperClass() {
        val model = JavaFileObjects
            .forResource("ModelWithAnnotatedClassAndSuperAttributes.java".patchResource())

        val generatedModel = JavaFileObjects
            .forResource("ModelWithAnnotatedClassAndSuperAttributes_.java".patchResource())

        val generatedSubClassModel = JavaFileObjects.forResource(
            (
                "ModelWithAnnotatedClassAndSuperAttributes\$SubModel" +
                    "WithAnnotatedClassAndSuperAttributes_.java"
                ).patchResource()
        )
        assertGeneration(
            sources = listOf(model),
            generatedFileObjects = listOf(generatedModel, generatedSubClassModel),
            // Throws NPE in KSP due to https://github.com/google/ksp/issues/622
            compilationMode = CompilationMode.JavaAP
        )
    }

    @Test
    fun testModelWithoutSetter() {
        assertGeneration("ModelWithoutSetter.java", "ModelWithoutSetter_.java")
    }

    @Test
    fun testModelReturningClassType() {
        assertGeneration("ModelReturningClassType.java", "ModelReturningClassType_.java")
    }

    @Test
    fun testModelReturningClassTypeWithVarargs() {
        assertGeneration(
            "ModelReturningClassTypeWithVarargs.java",
            "ModelReturningClassTypeWithVarargs_.java"
        )
    }

    @Test
    fun testModelWithVarargsConstructors() {
        assertGeneration(
            "ModelWithVarargsConstructors.java",
            "ModelWithVarargsConstructors_.java",
            // The kotlin extension generation doesn't actually work with vararg constructors.
            // it's a bug we're not trying to support, so just make sure this works with the originally
            // intended legacy java.
            compilationMode = CompilationMode.JavaAP
        )
    }

    @Test
    fun testModelWithHolderGeneratesNewHolderMethod() {
        assertGeneration(
            "AbstractModelWithHolder.java",
            "AbstractModelWithHolder_.java",
            // The nested static holder class results in KSP seeing it as an "Error" type during processing
            // for some reason (seems like a bug?). There doesn't seem to be a way to defer the
            // symbol the way we access it via a type param does not give the underlying KSAnnotated
            // element. The separate test with kotlin sources works fine, so workaround is to write
            // the holder classes in kotlin if needed.
            compilationMode = CompilationMode.JavaAP
        )
    }

    @Test
    fun testModelWithHolderGeneratesNewHolderMethod_kotlin() {
        assertGeneration(
            "testModelWithHolderGeneratesNewHolderMethod/AbstractModelWithHolder.kt",
            "testModelWithHolderGeneratesNewHolderMethod/AbstractModelWithHolder_.java",
            compilationMode = CompilationMode.KSP
        )
    }

    @Test
    fun testGenerateDefaultLayoutMethod() {
        assertGeneration("GenerateDefaultLayoutMethod.java", "GenerateDefaultLayoutMethod_.java")
    }

    @Test
    fun testGenerateDefaultLayoutMethodFailsIfLayoutNotSpecified() {
        assertGenerationError(
            "GenerateDefaultLayoutMethodNoLayout.java",
            "Model must specify a valid layout resource"
        )
    }

    @Test
    fun testGeneratedDefaultMethodWithLayoutSpecifiedInParent() {
        val model = JavaFileObjects
            .forResource("GenerateDefaultLayoutMethodParentLayout.java".patchResource())

        val generatedNoLayoutModel = JavaFileObjects
            .forResource("GenerateDefaultLayoutMethodParentLayout\$NoLayout_.java".patchResource())

        val generatedWithLayoutModel =
            JavaFileObjects.forResource("GenerateDefaultLayoutMethodParentLayout\$WithLayout_.java".patchResource())

        assertGeneration(
            sources = listOf(model),
            generatedFileObjects = listOf(generatedNoLayoutModel, generatedWithLayoutModel)
        )
    }

    @Test
    fun testGeneratedDefaultMethodWithLayoutSpecifiedInNextParent() {
        val model = JavaFileObjects
            .forResource("GenerateDefaultLayoutMethodNextParentLayout.java".patchResource())

        val generatedNoLayoutModel = JavaFileObjects
            .forResource("GenerateDefaultLayoutMethodNextParentLayout\$NoLayout_.java".patchResource())

        val generatedStillNoLayoutModel = JavaFileObjects
            .forResource("GenerateDefaultLayoutMethodNextParentLayout\$StillNoLayout_.java".patchResource())

        val generatedWithLayoutModel =
            JavaFileObjects.forResource("GenerateDefaultLayoutMethodNextParentLayout\$WithLayout_.java".patchResource())

        assertGeneration(
            sources = listOf(model),
            generatedFileObjects = listOf(
                generatedNoLayoutModel,
                generatedStillNoLayoutModel,
                generatedWithLayoutModel
            )
        )
    }

    @Test
    fun testGeneratedDefaultMethodWithLayoutFailsIfNotSpecifiedInParent() {
        assertGenerationError(
            "GenerateDefaultLayoutMethodParentStillNoLayout.java",
            "Model must specify a valid layout resource"
        )
    }

    @Test
    fun modelWithViewClickListener() {
        assertGeneration("ModelWithViewClickListener.java", "ModelWithViewClickListener_.java")
    }

    @Test
    fun modelWithViewClickLongListener() {
        val model = JavaFileObjects
            .forResource("ModelWithViewLongClickListener.java".patchResource())

        val generatedNoLayoutModel = JavaFileObjects
            .forResource("ModelWithViewLongClickListener_.java".patchResource())

        assertGeneration(
            sources = listOf(model),
            generatedFileObjects = listOf(generatedNoLayoutModel)
        )
    }

    @Test
    fun modelWithCheckedChangeListener() {
        assertGeneration(
            "ModelWithCheckedChangeListener.java",
            "ModelWithCheckedChangeListener_.java"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithoutGetterAndSetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithoutGetterAndSetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithoutSetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithoutSetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithoutGetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithoutGetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithIsPrefixGetter() {
        checkFileCompiles("ModelWithPrivateFieldWithIsPrefixGetter.java")
    }

    @Test
    fun testModelWithPrivateAttributeWithPrivateGetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithPrivateGetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithStaticGetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithStaticGetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithGetterWithParamsFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithGetterWithParams.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithPrivateSetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithPrivateSetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithStaticSetterFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithStaticSetter.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithPrivateAttributeWithGetterWithoutParamsFails() {
        assertGenerationError(
            "ModelWithPrivateFieldWithSettterWithoutParams.java",
            "private fields without proper getter and setter"
        )
    }

    @Test
    fun testModelWithAllPrivateFieldTypes() {
        assertGeneration(
            "ModelWithAllPrivateFieldTypes.java",
            "ModelWithAllPrivateFieldTypes_.java"
        )
    }

    @Test
    fun modelWithViewPrivateClickListener() {
        assertGeneration(
            "ModelWithPrivateViewClickListener.java",
            "ModelWithPrivateViewClickListener_.java"
        )
    }

    @Test
    fun modelWithPrivateFieldWithSameAsFieldGetterAndSetterName() {
        assertGeneration(
            "ModelWithPrivateFieldWithSameAsFieldGetterAndSetterName.java",
            "ModelWithPrivateFieldWithSameAsFieldGetterAndSetterName_.java"
        )
    }

    @Test
    fun testDoNotUseInToStringModel() {
        assertGeneration("ModelDoNotUseInToString.java", "ModelDoNotUseInToString_.java")
    }

    @Test
    fun modelWithAnnotation() {
        assertGeneration("ModelWithAnnotation.java", "ModelWithAnnotation_.java")
    }

    @Test
    fun testModelBuilderInterface() {
        assertGeneration("ModelWithAllFieldTypes.java", "ModelWithAllFieldTypesBuilder.java")
    }

    @Test
    fun generatedEpoxyModelGroup() {
        assertGeneration(
            "EpoxyModelGroupWithAnnotations.java",
            "EpoxyModelGroupWithAnnotations_.java",
        )
    }

    @Test
    fun generatedEpoxyModelWithView() {
        assertGeneration(
            "AbstractEpoxyModelWithView.java",
            "AbstractEpoxyModelWithView_.java"
        )
    }
}
